channel.c: defer exit_event

justinmk opened the pull request:
closes #8813
ref https://github.com/neovim/neovim/pull/7813#issuecomment-380949586
ref https://github.com/neovim/neovim/pull/8088#issuecomment-376312511

Deferring the event:

```diff
diff --git a/src/nvim/msgpack_rpc/channel.c b/src/nvim/msgpack_rpc/channel.c
index 3244d83a93c..62c7717c2e9 100644
--- a/src/nvim/msgpack_rpc/channel.c
+++ b/src/nvim/msgpack_rpc/channel.c
@@ -532,7 +532,7 @@ void rpc_close(Channel *channel)
   channel_decref(channel);
 
   if (channel->streamtype == kChannelStreamStdio) {
-    multiqueue_put(main_loop.fast_events, exit_event, 0);
+    multiqueue_put(main_loop.events, exit_event, 0);
   }
 }
```

...leads to hangs (waiting for input):
 
```
[5/7] Building C object test/functiona...s/CMakeFiles/tty-test.dir/tty-test.c.o[K
[6/7] Linking C executable bin/tty-test[K
[6/7] Linking C executable bin/tty-test[K
[7/7] cd /home/travis/build/neovim/neo...ild/neovim/neovim/cmake/RunTests.cmake[K

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

---

Notes from chat with @bfredl :

> The problem is that scheduling an event doesn't guarantee any progress back to the main normal mode.

* Screen tests may hang (regardless of success or failure), after the test harness (busted) closes the channel, but nvim is stuck in press-enter
* Root cause is that Nvim doesn't check for events during press-enter (`events_enabled=0` in `os/input.c`)
    * `normal_execute..wait_return..os_inchar` waits for input forever, because it doesn't process `main_loop.events`, because `events_enabled=0`, because our current policy is that it's unsafe to process `main_loop.events` until we get back to the top of the `state_enter` loop. See also https://github.com/neovim/neovim/pull/8814#issuecomment-422274556

justinmk committed:
channel.c: defer exit_event

closes #8813
ref https://github.com/neovim/neovim/pull/7813#issuecomment-380949586
ref https://github.com/neovim/neovim/pull/8088#issuecomment-376312511

Sometimes ASan fails like this:

    = ==16820==ERROR: AddressSanitizer: SEGV on unknown address 0x000001baab50 (pc 0x00000044f3a1 bp 0x000000000001 sp 0x7fff7ddb6a60 T0)
    ...
    4 0xfbfae4 in xfree /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:134:3
    5 0x120e948 in free_string_option /home/travis/build/neovim/neovim/build/../src/nvim/option.c:2207:5
    6 0x120de3d in free_all_options /home/travis/build/neovim/neovim/build/../src/nvim/option.c:923:9
    7 0xfc2441 in free_all_mem /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:622:3
    8 0x12bbe6c in mch_exit /home/travis/build/neovim/neovim/build/../src/nvim/os_unix.c:153:3
    9 0x1092cc3 in exit_event /home/travis/build/neovim/neovim/build/../src/nvim/msgpack_rpc/channel.c:542:5
    10 0xa6b2d3 in multiqueue_process_events /home/travis/build/neovim/neovim/build/../src/nvim/event/multiqueue.c:150:7
    11 0xa64e2f in loop_poll_events /home/travis/build/neovim/neovim/build/../src/nvim/event/loop.c:65:3
    12 0x12a4ec0 in os_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/os/input.c:162:5
    13 0x1035f07 in line_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/misc1.c:2692:5
    ...
    19 0x137ac07 in vim_regexec_nl /home/travis/build/neovim/neovim/build/../src/nvim/regexp.c:7302:10
    20 0x910f2b in f_split /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15673:17
    ...
    43 0x11667d5 in normal_check /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:1325:5
    44 0x17774c1 in state_enter /home/travis/build/neovim/neovim/build/../src/nvim/state.c:28:35
    45 0x10c0bf4 in normal_enter /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:467:3
    46 0xea98a2 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:566:3
    47 0x7fbb0a667f44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
    48 0x44e39b in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x44e39b)

Discussion:
- `free_all_options` is free'ing something before it was alloc'd.
    - ~~Options init isn't complete?~~ (No: init completed, `main.c`
    reached `normal_enter()`)
- stdio RPC channel was closed (so `rpc_close` schedules `exit_event`)
    - was this because the client died from a invalid message?
-  `exit_event` is scheduled on `main_loop.fast_events`.
    - would `main_loop.events` defer it? Or do we need something like
    `stuffReadbuff(":qall!\n")`?

Deferring the event:

    diff --git a/src/nvim/msgpack_rpc/channel.c b/src/nvim/msgpack_rpc/channel.c
    index 3244d83a93c..62c7717c2e9 100644
    --- a/src/nvim/msgpack_rpc/channel.c
    +++ b/src/nvim/msgpack_rpc/channel.c
    @@ -532,7 +532,7 @@ void rpc_close(Channel *channel)
       channel_decref(channel);

       if (channel->streamtype == kChannelStreamStdio) {
    -    multiqueue_put(main_loop.fast_events, exit_event, 0);
    +    multiqueue_put(main_loop.events, exit_event, 0);
       }
     }

...leads to hangs (waiting for input):

    [5/7] Building C object test/functiona...s/CMakeFiles/tty-test.dir/tty-test.c.o�[K
    [6/7] Linking C executable bin/tty-test�[K
    [6/7] Linking C executable bin/tty-test�[K
    [7/7] cd /home/travis/build/neovim/neo...ild/neovim/neovim/cmake/RunTests.cmake�[K

    No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
    Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

    The build has been terminated

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -539,7 +540,9 @@ void rpc_close(Channel *channel)
   if (!exiting) {
-    mch_exit(0);
+    // <C-\><C-N>:qall!<CR>
+    ins_typebuf((char_u *)"\034\016:qall!\n", REMAP_NONE, typebuf.tb_len, 1,
```
@bfredl Thanks for taking a look. Do you mean:

    multiqueue_put(main_loop.events, exit_event, 0);

that hangs (see PR description). But this `ins_typebuf()` also hangs, probably for the same reason.

bfredl replied:
Where does it hang? Possible to get a traceback?

justinmk replied:
workin' on it

bfredl commented:
Or should we just skip `free_all_mem` when nvim exits in a breakcheck? This is an artificial condition anyway not related to any real-world use.

bfredl commented:
Remember, vim is _by design_ not gc-safe in any place except for top-level, so I don't see how free_all_mem could be.

justinmk commented:
> Or should we just skip free_all_mem when nvim exits in a breakcheck?

But then we'd just get a different kind of false positive in the asan job, no?

bfredl commented:
I meant with some kind of message that a leak will be indeterministically (rarely) expected. A leak is a more reasonable failure than a crash that won't ever happen in real usage.

justinmk committed:
channel.c: defer exit_event

closes #8813
ref https://github.com/neovim/neovim/pull/7813#issuecomment-380949586
ref https://github.com/neovim/neovim/pull/8088#issuecomment-376312511

Sometimes ASan fails like this:

    = ==16820==ERROR: AddressSanitizer: SEGV on unknown address 0x000001baab50 (pc 0x00000044f3a1 bp 0x000000000001 sp 0x7fff7ddb6a60 T0)
    ...
    4 0xfbfae4 in xfree /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:134:3
    5 0x120e948 in free_string_option /home/travis/build/neovim/neovim/build/../src/nvim/option.c:2207:5
    6 0x120de3d in free_all_options /home/travis/build/neovim/neovim/build/../src/nvim/option.c:923:9
    7 0xfc2441 in free_all_mem /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:622:3
    8 0x12bbe6c in mch_exit /home/travis/build/neovim/neovim/build/../src/nvim/os_unix.c:153:3
    9 0x1092cc3 in exit_event /home/travis/build/neovim/neovim/build/../src/nvim/msgpack_rpc/channel.c:542:5
    10 0xa6b2d3 in multiqueue_process_events /home/travis/build/neovim/neovim/build/../src/nvim/event/multiqueue.c:150:7
    11 0xa64e2f in loop_poll_events /home/travis/build/neovim/neovim/build/../src/nvim/event/loop.c:65:3
    12 0x12a4ec0 in os_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/os/input.c:162:5
    13 0x1035f07 in line_breakcheck /home/travis/build/neovim/neovim/build/../src/nvim/misc1.c:2692:5
    ...
    19 0x137ac07 in vim_regexec_nl /home/travis/build/neovim/neovim/build/../src/nvim/regexp.c:7302:10
    20 0x910f2b in f_split /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15673:17
    ...
    43 0x11667d5 in normal_check /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:1325:5
    44 0x17774c1 in state_enter /home/travis/build/neovim/neovim/build/../src/nvim/state.c:28:35
    45 0x10c0bf4 in normal_enter /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:467:3
    46 0xea98a2 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:566:3
    47 0x7fbb0a667f44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
    48 0x44e39b in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x44e39b)

Discussion:
- `free_all_options` is free'ing something before it was alloc'd.
    - ~~Options init isn't complete?~~ (No: init completed, `main.c`
    reached `normal_enter()`)
- stdio RPC channel was closed (so `rpc_close` schedules `exit_event`)
    - was this because the client died from a invalid message?
-  `exit_event` is scheduled on `main_loop.fast_events`.
    - would `main_loop.events` defer it? Or do we need something like
    `stuffReadbuff(":qall!\n")`?

justinmk commented:
> Where does it hang? Possible to get a traceback?

It's stuck in `loop_poll_events`. And that makes sense, because `loop_poll_events` doesn't check `Loop.events`. But this also means that if we schedule something on `main_loop.events` it's not guaranteed to be processed until input (`Loop.uv`) is received.

Normally we don't want `loop_poll_events` to check `Loop.events` because that's application-specific. But I wonder if we should timeout `loop_poll_events` so that `normal_enter` occasionally checks `main_loop.events`.

```
(gdb) bt
+bt
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x000000000074fe52 in uv__io_poll (loop=<optimized out>, timeout=-1) at /home/vagrant/neovim/.deps/build/src/libuv/src/unix/linux-core.c:289
#2  0x0000000000742c64 in uv_run (loop=0xac5bd0 <main_loop>, mode=UV_RUN_ONCE) at /home/vagrant/neovim/.deps/build/src/libuv/src/unix/core.c:359
#3  0x00000000004ffbf3 in loop_poll_events (loop=0xac5bd0 <main_loop>, ms=-1) at ../src/nvim/event/loop.c:58
#4  0x0000000000627d7a in input_poll (ms=-1) at ../src/nvim/os/input.c:354
#5  0x0000000000626d90 in inbuf_poll (ms=-1) at ../src/nvim/os/input.c:377
#6  0x0000000000626cae in os_inchar (buf=0x0, maxlen=0, ms=-1, tb_change_cnt=0) at ../src/nvim/os/input.c:120
#7  0x00000000006cef1f in state_enter (s=0x7ffe4a3fe838) at ../src/nvim/state.c:55
#8  0x00000000005e4dc4 in normal_enter (cmdwin=false, noexmode=false) at ../src/nvim/normal.c:467
#9  0x00000000005a120f in main (argc=9, argv=0x7ffe4a3febd8) at ../src/nvim/main.c:566
```

bfredl commented:
That doesn't make sense and is a bug, ordinary normal mode should always process main_loop.events before waiting on input.

justinmk committed:
multiqueue.c: fix incorrect `size` counter

In multiqueue_push(), also increment the parent queue's size.

justinmk commented:
> That doesn't make sense and is a bug, ordinary normal mode should always process main_loop.events before waiting on input.

@bfredl  It does. But the stream might be closed _during_ the hit-enter prompt, after `events_enabled=0`. That's the problem: RPC events are not processed during hit-enter.

* We can deal with this in the test harness by sending CTRL-C or something like that before closing the stream. 
* Should we try a more robust solution which makes life easier for UI clients? 
    * Special-case the `exit_event` case? 
    * Process RPC events during press-enter?

bfredl commented:
> Process events during press-enter?

I would vote for this. clients shouldn't randomly do potentially dangerous stuff in an async context (i e not as part of handling an event from nvim) anyway, they can use `nvim_get_mode` and check the mode. Perhaps there should even be a mechanism for a client to ask "please send me a request as soon as nvim is back to top-level normal mode".

justinmk committed:
WIP: eventloop: Enable events "always"

If the stdio stream is closed, it's supposed to trigger
channel.c:exit_event. But if this happens (stream is closed) during
wait_return, it won't trigger (because events_enabled=0), and it waits
forever (for user input).

Example from system_spec.lua 'prints verbose information' test:

    DEBUG 2018-09-18T09:47:49.721 25418 UI: hl_attr_define
    DEBUG 2018-09-18T09:47:49.722 25418 pty_process_save_termios:47: tty_fd=-1
    DEBUG 2018-09-18T09:47:49.722 25418 rpc_start:71: rpc ch 1 in-stream=0x7fed14cad9c0 out-stream=0x7fed14cad820
    DEBUG 2018-09-18T09:47:49.725 25418 UI: hl_attr_define (+36 times...)
    DEBUG 2018-09-18T09:47:49.725 25418 UI: grid_clear
    DEBUG 2018-09-18T09:47:49.726 25418 receive_msgpack:226: ch 1: parsing 23 bytes from msgpack Stream: 0x7fed14cad820
    DEBUG 2018-09-18T09:47:49.726 25418 log_client_msg:732: RPC <-ch 1: [request]  [0, 0, "nvim_get_color_map", []]
    DEBUG 2018-09-18T09:47:49.726 25418 handle_request:354: RPC: scheduled nvim_get_color_map
    INFO  2018-09-18T09:47:49.726 25418 main:562: starting main loop
    DEBUG 2018-09-18T09:47:49.726 25418 UI: raw_line
    DEBUG 2018-09-18T09:47:49.726 25418 UI: raw_line (+42 times...)
    DEBUG 2018-09-18T09:47:49.726 25418 UI: grid_cursor_goto
    DEBUG 2018-09-18T09:47:49.726 25418 UI: mode_info_set
    DEBUG 2018-09-18T09:47:49.727 25418 UI: flush
    ERROR 2018-09-18T09:47:49.727 25418 input_poll:356: XXX blocking... events_enabled=0 empty=0
    DEBUG 2018-09-18T09:47:49.727 25418 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:106
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vpeekc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1546
      char_avail at /home/vagrant/neovim/build/../src/nvim/getchar.c:1573
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:40
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?
    DEBUG 2018-09-18T09:47:49.751 25418 state_enter:47: state: events
    DEBUG 2018-09-18T09:47:49.751 25418 state_enter:67: state: K_EVENT
    DEBUG 2018-09-18T09:47:49.751 25418 state_enter:71: input: 
    DEBUG 2018-09-18T09:47:49.751 25418 RPC: invoke nvim_get_color_map
    DEBUG 2018-09-18T09:47:49.751 25418 log_server_msg:695: RPC ->ch 1: [response] [1, 0, nil, {"AliceBlue"=>15792383, "AntiqueWhite"=>16444375, "AntiqueWhite1"=>16773083, "AntiqueWhite2"=>15654860, "AntiqueWhite3"=>13484208, "AntiqueWhite4"=>9143160, "Aqua"=>65535, "Aquamarine"=>8388564, "Aquamarine1"=>8388564, "Aquamarine2"=>7794374, "Aquamarine3"=>6737322, "Aquamarine4"=>4557684, "Azure"=>15794175, "Azure1"=>15794175, "Azure2"=>14741230, "Azure3"=>12701133, "Azure4"=>8620939, "Beige"=>16119260, "Bisque"=>16770244, "Bisque1"=>16770244, "Bisque2"=>15652279, "Bisque3"=>13481886, "Bisque4"=>9141611, "Black"=>0, "BlanchedAlmond"=>16772045, "Blue"=>255, "Blue1"=>255, "Blue2"=>238, "Blue3"=>205, "Blue4"=>139, "BlueViolet"=>9055202, "Brown"=>10824234, "Brown1"=>16728128, "Brown2"=>15612731, "Brown3"=>13447987, "Brown4"=>9118499, "BurlyWood"=>14596231, "Burlywood1"=>16765851, "Burlywood2"=>15648145, "Burlywood3"=>13478525, "Burlywood4"=>9139029, "CadetBlue"=>6266528, "CadetBlue1"=>10024447, "CadetBlue2"=>9364974, "CadetBlue3"=>8046029, "CadetBlue4"=>5473931, "ChartReuse"=>8388352, "Chartreuse1"=>8388352, "Chartreuse2"=>7794176, "Chartreuse3"=>6737152, "Chartreuse4"=>4557568, "Chocolate"=>13789470, "Chocolate1"=>16744228, "Chocolate2"=>15627809, "Chocolate3"=>13461021, "Chocolate4"=>9127187, "Coral"=>16744272, "Coral1"=>16740950, "Coral2"=>15624784, "Coral3"=>13458245, "Coral4"=>9125423, "CornFlowerBlue"=>6591981, "Cornsilk"=>16775388, "Cornsilk1"=>16775388, "Cornsilk2"=>15657165, "Cornsilk3"=>13486257, "Cornsilk4"=>9144440, "Crimson"=>14423100, "Cyan"=>65535, "Cyan1"=>65535, "Cyan2"=>61166, "Cyan3"=>52685, "Cyan4"=>35723, "DarkBlue"=>139, "DarkCyan"=>35723, "DarkGoldenRod"=>12092939, "DarkGoldenrod1"=>16759055, "DarkGoldenrod2"=>15641870, "DarkGoldenrod3"=>13473036, "DarkGoldenrod4"=>9135368, "DarkGray"=>11119017, "DarkGreen"=>25600, "DarkGrey"=>11119017, "DarkKhaki"=>12433259, "DarkMagenta"=>9109643, "DarkOliveGreen"=>5597999, "DarkOliveGreen1"=>13303664, "DarkOliveGreen2"=>12381800, "DarkOliveGreen3"=>10669402, "DarkOliveGreen4"=>7244605, "DarkOrange"=>16747520, "DarkOrange1"=>16744192, "DarkOrange2"=>15627776, "DarkOrange3"=>13460992, "DarkOrange4"=>9127168, "DarkOrchid"=>10040012, "DarkOrchid1"=>12533503, "DarkOrchid2"=>11680494, "DarkOrchid3"=>10105549, "DarkOrchid4"=>6824587, "DarkRed"=>9109504, "DarkSalmon"=>15308410, "DarkSeaGreen"=>9419919, "DarkSeaGreen1"=>12713921, "DarkSeaGreen2"=>11857588, "DarkSeaGreen3"=>10210715, "DarkSeaGreen4"=>6916969, "DarkSlateBlue"=>4734347, "DarkSlateGray"=>3100495, "DarkSlateGray1"=>9961471, "DarkSlateGray2"=>9301742, "DarkSlateGray3"=>7982541, "DarkSlateGray4"=>5409675, "DarkSlateGrey"=>3100495, "DarkTurquoise"=>52945, "DarkViolet"=>9699539, "DarkYellow"=>12303104, "DeepPink"=>16716947, "DeepPink1"=>16716947, "DeepPink2"=>15602313, "DeepPink3"=>13439094, "DeepPink4"=>9112144, "DeepSkyBlue"=>49151, "DeepSkyBlue1"=>49151, "DeepSkyBlue2"=>45806, "DeepSkyBlue3"=>39629, "DeepSkyBlue4"=>26763, "DimGray"=>6908265, "DimGrey"=>6908265, "DodgerBlue"=>2003199, "DodgerBlue1"=>2003199, "DodgerBlue2"=>1869550, "DodgerBlue3"=>1602765, "DodgerBlue4"=>1068683, "Firebrick"=>11674146, "Firebrick1"=>16724016, "Firebrick2"=>15608876, "Firebrick3"=>13444646, "Firebrick4"=>9116186, "FloralWhite"=>16775920, "ForestGreen"=>2263842, "Fuchsia"=>16711935, "Gainsboro"=>14474460, "GhostWhite"=>16316671, "Gold"=>16766720, "Gold1"=>16766720, "Gold2"=>15649024, "Gold3"=>13479168, "Gold4"=>9139456, "GoldenRod"=>14329120, "Goldenrod1"=>16761125, "Goldenrod2"=>15643682, "Goldenrod3"=>13474589, "Goldenrod4"=>9136404, "Gray"=>8421504, "Gray0"=>0, "Gray1"=>197379, "Gray10"=>1710618, "Gray100"=>16777215, "Gray11"=>1842204, "Gray12"=>2039583, "Gray13"=>2171169, "Gray14"=>2368548, "Gray15"=>2500134, "Gray16"=>2697513, "Gray17"=>2829099, "Gray18"=>3026478, "Gray19"=>3158064, "Gray2"=>328965, "Gray20"=>3355443, "Gray21"=>3552822, "Gray22"=>3684408, "Gray23"=>3881787, "Gray24"=>4013373, "Gray25"=>4210752, "Gray26"=>4342338, "Gray27"=>4539717, "Gray28"=>4671303, "Gray29"=>4868682, "Gray3"=>526344, "Gray30"=>5066061, "Gray31"=>5197647, "Gray32"=>5395026, "Gray33"=>5526612, "Gray34"=>5723991, "Gray35"=>5855577, "Gray36"=>6052956, "Gray37"=>6184542, "Gray38"=>6381921, "Gray39"=>6513507, "Gray4"=>657930, "Gray40"=>6710886, "Gray41"=>6908265, "Gray42"=>7039851, "Gray43"=>7237230, "Gray44"=>7368816, "Gray45"=>7566195, "Gray46"=>7697781, "Gray47"=>7895160, "Gray48"=>8026746, "Gray49"=>8224125, "Gray5"=>855309, "Gray50"=>8355711, "Gray51"=>8553090, "Gray52"=>8750469, "Gray53"=>8882055, "Gray54"=>9079434, "Gray55"=>9211020, "Gray56"=>9408399, "Gray57"=>9539985, "Gray58"=>9737364, "Gray59"=>9868950, "Gray6"=>986895, "Gray60"=>10066329, "Gray61"=>10263708, "Gray62"=>10395294, "Gray63"=>10592673, "Gray64"=>10724259, "Gray65"=>10921638, "Gray66"=>11053224, "Gray67"=>11250603, "Gray68"=>11382189, "Gray69"=>11579568, "Gray7"=>1184274, "Gray70"=>11776947, "Gray71"=>11908533, "Gray72"=>12105912, "Gray73"=>12237498, "Gray74"=>12434877, "Gray75"=>12566463, "Gray76"=>12763842, "Gray77"=>12895428, "Gray78"=>13092807, "Gray79"=>13224393, "Gray8"=>1315860, "Gray80"=>13421772, "Gray81"=>13619151, "Gray82"=>13750737, "Gray83"=>13948116, "Gray84"=>14079702, "Gray85"=>14277081, "Gray86"=>14408667, "Gray87"=>14606046, "Gray88"=>14737632, "Gray89"=>14935011, "Gray9"=>1513239, "Gray90"=>15066597, "Gray91"=>15263976, "Gray92"=>15461355, "Gray93"=>15592941, "Gray94"=>15790320, "Gray95"=>15921906, "Gray96"=>16119285, "Gray97"=>16250871, "Gray98"=>16448250, "Gray99"=>16579836, "Green"=>32768, "Green1"=>65280, "Green2"=>60928, "Green3"=>52480, "Green4"=>35584, "GreenYellow"=>11403055, "Grey"=>8421504, "Grey0"=>0, "Grey1"=>197379, "Grey10"=>1710618, "Grey100"=>16777215, "Grey11"=>1842204, "Grey12"=>2039583, "Grey13"=>2171169, "Grey14"=>2368548, "Grey15"=>2500134, "Grey16"=>2697513, "Grey17"=>2829099, "Grey18"=>3026478, "Grey19"=>3158064, "Grey2"=>328965, "Grey20"=>3355443, "Grey21"=>3552822, "Grey22"=>3684408, "Grey23"=>3881787, "Grey24"=>4013373, "Grey25"=>4210752, "Grey26"=>4342338, "Grey27"=>4539717, "Grey28"=>4671303, "Grey29"=>4868682, "Grey3"=>526344, "Grey30"=>5066061, "Grey31"=>5197647, "Grey32"=>5395026, "Grey33"=>5526612, "Grey34"=>5723991, "Grey35"=>5855577, "Grey36"=>6052956, "Grey37"=>6184542, "Grey38"=>6381921, "Grey39"=>6513507, "Grey4"=>657930, "Grey40"=>6710886, "Grey41"=>6908265, "Grey42"=>7039851, "Grey43"=>7237230, "Grey44"=>7368816, "Grey45"=>7566195, "Grey46"=>7697781, "Grey47"=>7895160, "Grey48"=>8026746, "Grey49"=>8224125, "Grey5"=>855309, "Grey50"=>8355711, "Grey51"=>8553090, "Grey52"=>8750469, "Grey53"=>8882055, "Grey54"=>9079434, "Grey55"=>9211020, "Grey56"=>9408399, "Grey57"=>9539985, "Grey58"=>9737364, "Grey59"=>9868950, "Grey6"=>986895, "Grey60"=>10066329, "Grey61"=>10263708, "Grey62"=>10395294, "Grey63"=>10592673, "Grey64"=>10724259, "Grey65"=>10921638, "Grey66"=>11053224, "Grey67"=>11250603, "Grey68"=>11382189, "Grey69"=>11579568, "Grey7"=>1184274, "Grey70"=>11776947, "Grey71"=>11908533, "Grey72"=>12105912, "Grey73"=>12237498, "Grey74"=>12434877, "Grey75"=>12566463, "Grey76"=>12763842, "Grey77"=>12895428, "Grey78"=>13092807, "Grey79"=>13224393, "Grey8"=>1315860, "Grey80"=>13421772, "Grey81"=>13619151, "Grey82"=>13750737, "Grey83"=>13948116, "Grey84"=>14079702, "Grey85"=>14277081, "Grey86"=>14408667, "Grey87"=>14606046, "Grey88"=>14737632, "Grey89"=>14935011, "Grey9"=>1513239, "Grey90"=>15066597, "Grey91"=>15263976, "Grey92"=>15461355, "Grey93"=>15592941, "Grey94"=>15790320, "Grey95"=>15921906, "Grey96"=>16119285, "Grey97"=>16250871, "Grey98"=>16448250, "Grey99"=>16579836, "Honeydew"=>15794160, "Honeydew1"=>15794160, "Honeydew2"=>14741216, "Honeydew3"=>12701121, "Honeydew4"=>8620931, "HotPink"=>16738740, "HotPink1"=>16740020, "HotPink2"=>15624871, "HotPink3"=>13459600, "HotPink4"=>9124450, "IndianRed"=>13458524, "IndianRed1"=>16738922, "IndianRed2"=>15623011, "IndianRed3"=>13456725, "IndianRed4"=>9124410, "Indigo"=>4915330, "Ivory"=>16777200, "Ivory1"=>16777200, "Ivory2"=>15658720, "Ivory3"=>13487553, "Ivory4"=>9145219, "Khaki"=>15787660, "Khaki1"=>16774799, "Khaki2"=>15656581, "Khaki3"=>13485683, "Khaki4"=>9143886, "Lavender"=>15132410, "LavenderBlush"=>16773365, "LavenderBlush1"=>16773365, "LavenderBlush2"=>15655141, "LavenderBlush3"=>13484485, "LavenderBlush4"=>9143174, "LawnGreen"=>8190976, "LemonChiffon"=>16775885, "LemonChiffon1"=>16775885, "LemonChiffon2"=>15657407, "LemonChiffon3"=>13486501, "LemonChiffon4"=>9144688, "LightBlue"=>11393254, "LightBlue1"=>12578815, "LightBlue2"=>11722734, "LightBlue3"=>10141901, "LightBlue4"=>6849419, "LightCoral"=>15761536, "LightCyan"=>14745599, "LightCyan1"=>14745599, "LightCyan2"=>13758190, "LightCyan3"=>11849165, "LightCyan4"=>8031115, "LightGoldenrod"=>15654274, "LightGoldenrod1"=>16772235, "LightGoldenrod2"=>15654018, "LightGoldenrod3"=>13483632, "LightGoldenrod4"=>9142604, "LightGoldenRodYellow"=>16448210, "LightGray"=>13882323, "LightGreen"=>9498256, "LightGrey"=>13882323, "LightMagenta"=>16759807, "LightPink"=>16758465, "LightPink1"=>16756409, "LightPink2"=>15639213, "LightPink3"=>13470869, "LightPink4"=>9133925, "LightRed"=>16759739, "LightSalmon"=>16752762, "LightSalmon1"=>16752762, "LightSalmon2"=>15635826, "LightSalmon3"=>13468002, "LightSalmon4"=>9131842, "LightSeaGreen"=>2142890, "LightSkyBlue"=>8900346, "LightSkyBlue1"=>11592447, "LightSkyBlue2"=>10802158, "LightSkyBlue3"=>9287373, "LightSkyBlue4"=>6323083, "LightSlateBlue"=>8679679, "LightSlateGray"=>7833753, "LightSlateGrey"=>7833753, "LightSteelBlue"=>11584734, "LightSteelBlue1"=>13296127, "LightSteelBlue2"=>12374766, "LightSteelBlue3"=>10663373, "LightSteelBlue4"=>7240587, "LightYellow"=>16777184, "LightYellow1"=>16777184, "LightYellow2"=>15658705, "LightYellow3"=>13487540, "LightYellow4"=>9145210, "Lime"=>65280, "LimeGreen"=>3329330, "Linen"=>16445670, "Magenta"=>16711935, "Magenta1"=>16711935, "Magenta2"=>15597806, "Magenta3"=>13435085, "Magenta4"=>9109643, "Maroon"=>8388608, "Maroon1"=>16725171, "Maroon2"=>15610023, "Maroon3"=>13445520, "Maroon4"=>9116770, "MediumAquamarine"=>6737322, "MediumBlue"=>205, "MediumOrchid"=>12211667, "MediumOrchid1"=>14706431, "MediumOrchid2"=>13721582, "MediumOrchid3"=>11817677, "MediumOrchid4"=>8009611, "MediumPurple"=>9662683, "MediumPurple1"=>11240191, "MediumPurple2"=>10451438, "MediumPurple3"=>9005261, "MediumPurple4"=>6113163, "MediumSeaGreen"=>3978097, "MediumSlateBlue"=>8087790, "MediumSpringGreen"=>64154, "MediumTurquoise"=>4772300, "MediumVioletRed"=>13047173, "MidnightBlue"=>1644912, "MintCream"=>16121850, "MistyRose"=>16770273, "MistyRose1"=>16770273, "MistyRose2"=>15652306, "MistyRose3"=>13481909, "MistyRose4"=>9141627, "Moccasin"=>16770229, "NavajoWhite"=>16768685, "NavajoWhite1"=>16768685, "NavajoWhite2"=>15650721, "NavajoWhite3"=>13480843, "NavajoWhite4"=>9140574, "Navy"=>128, "NavyBlue"=>128, "OldLace"=>16643558, "Olive"=>8421376, "OliveDrab"=>7048739, "OliveDrab1"=>12648254, "OliveDrab2"=>11791930, "OliveDrab3"=>10145074, "OliveDrab4"=>6916898, "Orange"=>16753920, "Orange1"=>16753920, "Orange2"=>15636992, "Orange3"=>13468928, "Orange4"=>9132544, "OrangeRed"=>16729344, "OrangeRed1"=>16729344, "OrangeRed2"=>15613952, "OrangeRed3"=>13448960, "OrangeRed4"=>9118976, "Orchid"=>14315734, "Orchid1"=>16745466, "Orchid2"=>15629033, "Orchid3"=>13461961, "Orchid4"=>9127817, "PaleGoldenRod"=>15657130, "PaleGreen"=>10025880, "PaleGreen1"=>10157978, "PaleGreen2"=>9498256, "PaleGreen3"=>8179068, "PaleGreen4"=>5540692, "PaleTurquoise"=>11529966, "PaleTurquoise1"=>12320767, "PaleTurquoise2"=>11464430, "PaleTurquoise3"=>9883085, "PaleTurquoise4"=>6720395, "PaleVioletRed"=>14381203, "PaleVioletRed1"=>16745131, "PaleVioletRed2"=>15628703, "PaleVioletRed3"=>13461641, "PaleVioletRed4"=>9127773, "PapayaWhip"=>16773077, "PeachPuff"=>16767673, "PeachPuff1"=>16767673, "PeachPuff2"=>15649709, "PeachPuff3"=>13479829, "PeachPuff4"=>9140069, "Peru"=>13468991, "Pink"=>16761035, "Pink1"=>16758213, "Pink2"=>15641016, "Pink3"=>13472158, "Pink4"=>9134956, "Plum"=>14524637, "Plum1"=>16759807, "Plum2"=>15642350, "Plum3"=>13473485, "Plum4"=>9135755, "PowderBlue"=>11591910, "Purple"=>8388736, "Purple1"=>10170623, "Purple2"=>9514222, "Purple3"=>8201933, "Purple4"=>5577355, "RebeccaPurple"=>6697881, "Red"=>16711680, "Red1"=>16711680, "Red2"=>15597568, "Red3"=>13434880, "Red4"=>9109504, "RosyBrown"=>12357519, "RosyBrown1"=>16761281, "RosyBrown2"=>15643828, "RosyBrown3"=>13474715, "RosyBrown4"=>9136489, "RoyalBlue"=>4286945, "RoyalBlue1"=>4749055, "RoyalBlue2"=>4419310, "RoyalBlue3"=>3825613, "RoyalBlue4"=>2572427, "SaddleBrown"=>9127187, "Salmon"=>16416882, "Salmon1"=>16747625, "Salmon2"=>15630946, "Salmon3"=>13463636, "Salmon4"=>9129017, "SandyBrown"=>16032864, "SeaGreen"=>3050327, "SeaGreen1"=>5570463, "SeaGreen2"=>5172884, "SeaGreen3"=>4443520, "SeaGreen4"=>3050327, "SeaShell"=>16774638, "Seashell1"=>16774638, "Seashell2"=>15656414, "Seashell3"=>13485503, "Seashell4"=>9143938, "Sienna"=>10506797, "Sienna1"=>16745031, "Sienna2"=>15628610, "Sienna3"=>13461561, "Sienna4"=>9127718, "Silver"=>12632256, "SkyBlue"=>8900331, "SkyBlue1"=>8900351, "SkyBlue2"=>8306926, "SkyBlue3"=>7120589, "SkyBlue4"=>4878475, "SlateBlue"=>6970061, "SlateBlue1"=>8613887, "SlateBlue2"=>8021998, "SlateBlue3"=>6904269, "SlateBlue4"=>4668555, "SlateGray"=>7372944, "SlateGray1"=>13034239, "SlateGray2"=>12178414, "SlateGray3"=>10467021, "SlateGray4"=>7109515, "SlateGrey"=>7372944, "Snow"=>16775930, "Snow1"=>16775930, "Snow2"=>15657449, "Snow3"=>13486537, "Snow4"=>9144713, "SpringGreen"=>65407, "SpringGreen1"=>65407, "SpringGreen2"=>61046, "SpringGreen3"=>52582, "SpringGreen4"=>35653, "SteelBlue"=>4620980, "SteelBlue1"=>6535423, "SteelBlue2"=>6073582, "SteelBlue3"=>5215437, "SteelBlue4"=>3564683, "Tan"=>13808780, "Tan1"=>16753999, "Tan2"=>15637065, "Tan3"=>13468991, "Tan4"=>9132587, "Teal"=>32896, "Thistle"=>14204888, "Thistle1"=>16769535, "Thistle2"=>15651566, "Thistle3"=>13481421, "Thistle4"=>9141131, "Tomato"=>16737095, "Tomato1"=>16737095, "Tomato2"=>15621186, "Tomato3"=>13455161, "Tomato4"=>9123366, "Turquoise"=>4251856, "Turquoise1"=>62975, "Turquoise2"=>58862, "Turquoise3"=>50637, "Turquoise4"=>34443, "Violet"=>15631086, "VioletRed"=>13639824, "VioletRed1"=>16727702, "VioletRed2"=>15612556, "VioletRed3"=>13447800, "VioletRed4"=>9118290, "WebGray"=>8421504, "WebGreen"=>32768, "WebGrey"=>8421504, "WebMaroon"=>8388608, "WebPurple"=>8388736, "Wheat"=>16113331, "Wheat1"=>16771002, "Wheat2"=>15653038, "Wheat3"=>13482646, "Wheat4"=>9141862, "White"=>16777215, "WhiteSmoke"=>16119285, "X11Gray"=>12500670, "X11Green"=>65280, "X11Grey"=>12500670, "X11Maroon"=>11546720, "X11Purple"=>10494192, "Yellow"=>16776960, "Yellow1"=>16776960, "Yellow2"=>15658496, "Yellow3"=>13487360, "Yellow4"=>9145088, "YellowGreen"=>10145074}]
    ERROR 2018-09-18T09:47:49.753 25418 input_poll:356: XXX blocking... events_enabled=0 empty=1
    DEBUG 2018-09-18T09:47:49.753 25418 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:106
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vpeekc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1546
      char_avail at /home/vagrant/neovim/build/../src/nvim/getchar.c:1573
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:40
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?
    DEBUG 2018-09-18T09:47:49.796 25418 read_cb:122: Closing Stream (0x7fed14cad820): EOF (end of file)
    DEBUG 2018-09-18T09:47:49.796 25418 stream_close:85: closing Stream: 0x7fed14cad820
    DEBUG 2018-09-18T09:47:49.796 25418 stream_close:85: closing Stream: 0x7fed14cad9c0
    WARN  2018-09-18T09:47:49.796 25418 call_set_error:605: RPC: ch 1 was closed by the client
    DEBUG 2018-09-18T09:47:49.796 25418 UI: flush (+2 times...)
    DEBUG 2018-09-18T09:47:49.796 25418 UI: stop
    ERROR 2018-09-18T09:47:49.797 25418 mch_exit:156: XXX exited
    DEBUG 2018-09-18T09:47:49.805 25423 UI: hl_attr_define
    DEBUG 2018-09-18T09:47:49.806 25423 pty_process_save_termios:47: tty_fd=-1
    DEBUG 2018-09-18T09:47:49.808 25423 rpc_start:71: rpc ch 1 in-stream=0x7ffb8caad9c0 out-stream=0x7ffb8caad820
    DEBUG 2018-09-18T09:47:49.812 25423 UI: hl_attr_define (+36 times...)
    DEBUG 2018-09-18T09:47:49.812 25423 UI: grid_clear
    DEBUG 2018-09-18T09:47:49.813 25423 receive_msgpack:226: ch 1: parsing 40 bytes from msgpack Stream: 0x7ffb8caad820
    DEBUG 2018-09-18T09:47:49.813 25423 log_client_msg:732: RPC <-ch 1: [request]  [0, 0, "nvim_ui_attach", [53, 14, {"rgb"=>true, "ext_newgrid"=>true}]]
    DEBUG 2018-09-18T09:47:49.814 25423 handle_request:354: RPC: scheduled nvim_ui_attach
    INFO  2018-09-18T09:47:49.814 25423 main:562: starting main loop
    DEBUG 2018-09-18T09:47:49.815 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.815 25423 UI: raw_line (+21 times...)
    DEBUG 2018-09-18T09:47:49.816 25423 UI: mode_info_set
    DEBUG 2018-09-18T09:47:49.816 25423 UI: flush
    ERROR 2018-09-18T09:47:49.816 25423 input_poll:356: XXX blocking... events_enabled=0 empty=0
    DEBUG 2018-09-18T09:47:49.817 25423 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:106
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vpeekc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1546
      char_avail at /home/vagrant/neovim/build/../src/nvim/getchar.c:1573
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:40
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?
    DEBUG 2018-09-18T09:47:49.839 25423 state_enter:47: state: events
    DEBUG 2018-09-18T09:47:49.839 25423 state_enter:67: state: K_EVENT
    DEBUG 2018-09-18T09:47:49.839 25423 state_enter:71: input: 
    DEBUG 2018-09-18T09:47:49.840 25423 RPC: invoke nvim_ui_attach
    DEBUG 2018-09-18T09:47:49.840 25423 UI: option_set
    DEBUG 2018-09-18T09:47:49.840 25423 UI: option_set (+8 times...)
    DEBUG 2018-09-18T09:47:49.840 25423 UI: default_colors_set
    DEBUG 2018-09-18T09:47:49.840 25423 UI: grid_resize
    DEBUG 2018-09-18T09:47:49.840 25423 UI: grid_clear
    DEBUG 2018-09-18T09:47:49.840 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.840 25423 UI: raw_line (+11 times...)
    DEBUG 2018-09-18T09:47:49.840 25423 UI: grid_cursor_goto
    DEBUG 2018-09-18T09:47:49.840 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.840 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["option_set", ["arabicshape", true], ["ambiwidth", "single"], ["emoji", true], ["guifont", ""], ["guifontset", ""], ["guifontwide", ""], ["linespace", 0], ["showtabline", 1], ["termguicolors", false], ["ext_newgrid", true], ["ext_hlstate", false]], ["hl_attr_define", [1, {}, {}, []], [2, {"foreground"=>13882323, "background"=>11119017}, {"foreground"=>7, "background"=>242}, []], [3, {"bold"=>true, "background"=>16711680}, {"bold"=>true, "background"=>9}, []], [4, {"foreground"=>16777215, "background"=>16711680}, {"foreground"=>15, "background"=>1}, []], [5, {"reverse"=>true}, {"reverse"=>true}, []], [6, {"bold"=>true}, {"bold"=>true}, []], [7, {"bold"=>true, "foreground"=>255}, {"foreground"=>12}, []], [8, {"background"=>8421504}, {"background"=>248}, []], [9, {"bold"=>true, "reverse"=>true}, {"bold"=>true, "reverse"=>true}, []], [10, {"foreground"=>0, "background"=>16776960}, {"foreground"=>0, "background"=>11}, []], [11, {"background"=>16759739}, {"background"=>224}, []], [12, {"background"=>15066597}, {"background"=>7}, []], [13, {"background"=>15066597}, {"underline"=>true}, []], [14, {"bold"=>true, "foreground"=>10824234}, {"foreground"=>130}, []], [15, {"background"=>11393254}, {"background"=>81}, []], [16, {"background"=>16759807}, {"background"=>225}, []], [17, {"bold"=>true, "foreground"=>255, "background"=>14745599}, {"foreground"=>12, "background"=>159}, []], [18, {"foreground"=>255}, {"foreground"=>4}, []], [19, {"foreground"=>139, "background"=>8421504}, {"foreground"=>4, "background"=>248}, []], [20, {"foreground"=>139, "background"=>13882323}, {"foreground"=>4, "background"=>248}, []], [21, {"foreground"=>10824234}, {"foreground"=>130}, []], [22, {"background"=>65535}, {"background"=>14}, []], [23, {"bold"=>true, "foreground"=>3050327}, {"foreground"=>2}, []], [24, {"background"=>16759807}, {"foreground"=>0, "background"=>225}, []], [25, {"background"=>8421504}, {"foreground"=>0, "background"=>7}, []], [26, {"background"=>0}, {"background"=>0}, []], [27, {"background"=>16776960}, {"background"=>11}, []], [28, {"undercurl"=>true, "special"=>16711680}, {"background"=>224}, []], [29, {"undercurl"=>true, "special"=>255}, {"background"=>81}, []], [30, {"undercurl"=>true, "special"=>35723}, {"background"=>14}, []], [31, {"undercurl"=>true, "special"=>16711935}, {"background"=>225}, []], [32, {"underline"=>true, "background"=>13882323}, {"underline"=>true, "foreground"=>0, "background"=>7}, []], [33, {"bold"=>true, "foreground"=>16711935}, {"foreground"=>5}, []], [34, {"background"=>13882323}, {}, []], [35, {"foreground"=>16711680}, {"foreground"=>1}, []], [36, {"background"=>13882323}, {"background"=>7}, []], [37, {"foreground"=>16711680, "background"=>16711680}, {"foreground"=>9, "background"=>9}, []]], ["default_colors_set", [-1, -1, -1, 0, 0]], ["grid_resize", [1, 53, 14]], ["grid_clear", [1]], ["grid_line", [1, 1, 0, [["~", 7], [" ", 7, 52]]], [1, 2, 0, [["~", 7], [" ", 7, 52]]], [1, 3, 0, [["~", 7], [" ", 7, 52]]], [1, 4, 0, [["~", 7], [" ", 7, 52]]], [1, 5, 0, [["~", 7], [" ", 7, 52]]], [1, 6, 0, [["~", 7], [" ", 7, 52]]], [1, 7, 0, [["~", 7], [" ", 7, 52]]], [1, 8, 0, [["~", 7], [" ", 7, 52]]], [1, 9, 0, [["~", 7], [" ", 7, 52]]], [1, 10, 0, [["~", 7], [" ", 7, 52]]], [1, 11, 0, [["~", 7], [" ", 7, 52]]], [1, 12, 0, [["~", 7], [" ", 7, 52]]]], ["grid_cursor_goto", [1, 0, 0]]]]
    DEBUG 2018-09-18T09:47:49.840 25423 UI: option_set
    DEBUG 2018-09-18T09:47:49.841 25423 log_server_msg:695: RPC ->ch 1: [response] [1, 0, nil, nil]
    DEBUG 2018-09-18T09:47:49.841 25423 UI: option_set (+3 times...)
    DEBUG 2018-09-18T09:47:49.841 25423 UI: mode_info_set
    DEBUG 2018-09-18T09:47:49.841 25423 UI: mode_change
    DEBUG 2018-09-18T09:47:49.841 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.841 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["option_set", ["ext_cmdline", false], ["ext_popupmenu", false], ["ext_tabline", false], ["ext_wildmenu", false]], ["mode_info_set", [true, [{"mouse_shape"=>0, "cursor_shape"=>"block", "cell_percentage"=>0, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"normal", "short_name"=>"n"}, {"mouse_shape"=>0, "cursor_shape"=>"block", "cell_percentage"=>0, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"visual", "short_name"=>"v"}, {"mouse_shape"=>0, "cursor_shape"=>"vertical", "cell_percentage"=>25, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"insert", "short_name"=>"i"}, {"mouse_shape"=>0, "cursor_shape"=>"horizontal", "cell_percentage"=>20, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"replace", "short_name"=>"r"}, {"mouse_shape"=>0, "cursor_shape"=>"block", "cell_percentage"=>0, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"cmdline_normal", "short_name"=>"c"}, {"mouse_shape"=>0, "cursor_shape"=>"vertical", "cell_percentage"=>25, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"cmdline_insert", "short_name"=>"ci"}, {"mouse_shape"=>0, "cursor_shape"=>"horizontal", "cell_percentage"=>20, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"cmdline_replace", "short_name"=>"cr"}, {"mouse_shape"=>0, "cursor_shape"=>"horizontal", "cell_percentage"=>20, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"operator", "short_name"=>"o"}, {"mouse_shape"=>0, "cursor_shape"=>"vertical", "cell_percentage"=>25, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"visual_select", "short_name"=>"ve"}, {"mouse_shape"=>0, "name"=>"cmdline_hover", "short_name"=>"e"}, {"mouse_shape"=>0, "name"=>"statusline_hover", "short_name"=>"s"}, {"mouse_shape"=>0, "name"=>"statusline_drag", "short_name"=>"sd"}, {"mouse_shape"=>0, "name"=>"vsep_hover", "short_name"=>"vs"}, {"mouse_shape"=>0, "name"=>"vsep_drag", "short_name"=>"vd"}, {"mouse_shape"=>0, "name"=>"more", "short_name"=>"m"}, {"mouse_shape"=>0, "name"=>"more_lastline", "short_name"=>"ml"}, {"cursor_shape"=>"block", "cell_percentage"=>0, "blinkwait"=>0, "blinkon"=>0, "blinkoff"=>0, "hl_id"=>0, "id_lm"=>0, "attr_id"=>0, "attr_id_lm"=>0, "name"=>"showmatch", "short_name"=>"sm"}]]], ["mode_change", ["normal", 0]]]]
    ERROR 2018-09-18T09:47:49.841 25423 input_poll:356: XXX blocking... events_enabled=0 empty=1
    DEBUG 2018-09-18T09:47:49.841 25423 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:106
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vpeekc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1546
      char_avail at /home/vagrant/neovim/build/../src/nvim/getchar.c:1573
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:40
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?
    DEBUG 2018-09-18T09:47:49.871 25423 receive_msgpack:226: ch 1: parsing 60 bytes from msgpack Stream: 0x7ffb8caad820
    DEBUG 2018-09-18T09:47:49.871 25423 log_client_msg:732: RPC <-ch 1: [request]  [0, 1, "nvim_input", [":4verbose echo system("echo ".getpid())<cr>"]]
    DEBUG 2018-09-18T09:47:49.871 25423 RPC: invoke nvim_input
    DEBUG 2018-09-18T09:47:49.871 25423 log_server_msg:695: RPC ->ch 1: [response] [1, 1, nil, 43]
    DEBUG 2018-09-18T09:47:49.871 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.872 25423 state_enter:71: input: :
    DEBUG 2018-09-18T09:47:49.872 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:71: input: 4
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:71: input: v
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:71: input: r
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.873 25423 state_enter:71: input: b
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: o
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: s
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input:
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: c
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: h
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input: o
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.874 25423 state_enter:71: input:
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: s
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: y
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: s
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: t
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: m
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: (
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: "
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: c
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: h
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.875 25423 state_enter:71: input: o
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input:
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: "
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: .
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: g
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: e
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: t
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: p
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: i
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: d
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:71: input: (
    DEBUG 2018-09-18T09:47:49.876 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.877 25423 state_enter:71: input: )
    DEBUG 2018-09-18T09:47:49.880 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.880 25423 state_enter:71: input: )
    DEBUG 2018-09-18T09:47:49.880 25423 state_enter:41: state: avail
    DEBUG 2018-09-18T09:47:49.881 25423 state_enter:71: input:
    DEBUG 2018-09-18T09:47:49.881 25423 UI: raw_line (+35 times...)
    DEBUG 2018-09-18T09:47:49.881 25423 UI: grid_cursor_goto
    DEBUG 2018-09-18T09:47:49.881 25423 UI: mode_change
    DEBUG 2018-09-18T09:47:49.881 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.881 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["grid_line", [1, 13, 0, [[":", 0]]], [1, 13, 1, [["4", 0]]], [1, 13, 2, [["v", 0]]], [1, 13, 3, [["e", 0]]], [1, 13, 4, [["r", 0]]], [1, 13, 5, [["b", 0]]], [1, 13, 6, [["o", 0]]], [1, 13, 7, [["s", 0]]], [1, 13, 8, [["e", 0]]], [1, 13, 10, [["e", 0]]], [1, 13, 11, [["c", 0]]], [1, 13, 12, [["h", 0]]], [1, 13, 13, [["o", 0]]], [1, 13, 15, [["s", 0]]], [1, 13, 16, [["y", 0]]], [1, 13, 17, [["s", 0]]], [1, 13, 18, [["t", 0]]], [1, 13, 19, [["e", 0]]], [1, 13, 20, [["m", 0]]], [1, 13, 21, [["(", 0]]], [1, 13, 22, [[""", 0]]], [1, 13, 23, [["e", 0]]], [1, 13, 24, [["c", 0]]], [1, 13, 25, [["h", 0]]], [1, 13, 26, [["o", 0]]], [1, 13, 28, [[""", 0]]], [1, 13, 29, [[".", 0]]], [1, 13, 30, [["g", 0]]], [1, 13, 31, [["e", 0]]], [1, 13, 32, [["t", 0]]], [1, 13, 33, [["p", 0]]], [1, 13, 34, [["i", 0]]], [1, 13, 35, [["d", 0]]], [1, 13, 36, [["(", 0]]], [1, 13, 37, [[")", 0]]], [1, 13, 38, [[")", 0]]]], ["grid_cursor_goto", [1, 13, 0]], ["mode_change", ["cmdline_normal", 4]]]]
    DEBUG 2018-09-18T09:47:49.882 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.882 25423 UI: raw_line (+1 times...)
    DEBUG 2018-09-18T09:47:49.882 25423 UI: grid_cursor_goto
    DEBUG 2018-09-18T09:47:49.883 25423 UI: mode_change
    DEBUG 2018-09-18T09:47:49.883 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.883 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["grid_line", [1, 13, 0, [[" ", 0, 39]]], [1, 13, 0, [["C", 0], ["a"], ["l", 0, 2], ["i"], ["n"], ["g"], [" "], ["s"], ["h"], ["e"], ["l", 0, 2], [" "], ["t"], ["o"], [" "], ["e"], ["x"], ["e"], ["c"], ["u"], ["t"], ["e"], [":"], [" "], ["""], ["e"], ["c"], ["h"], ["o"], [" "], ["2"], ["5"], ["4"], ["2"], ["3"], ["""]]]], ["grid_cursor_goto", [1, 13, 38]], ["mode_change", ["normal", 0]]]]
    DEBUG 2018-09-18T09:47:49.884 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.884 25423 UI: grid_scroll
    DEBUG 2018-09-18T09:47:49.888 25423 process_spawn:114: new: pid=25428 argv=[/bin/bash]
    DEBUG 2018-09-18T09:47:49.888 25423 UI: grid_scroll (+1 times...)
    DEBUG 2018-09-18T09:47:49.888 25423 UI: busy_start
    DEBUG 2018-09-18T09:47:49.888 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.888 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["grid_line", [1, 12, 0, [[" ", 9, 53]]]], ["grid_scroll", [1, 11, 14, 0, 53, 1, 0], [1, 10, 14, 0, 53, 1, 0]], ["busy_start", []]]]
    DEBUG 2018-09-18T09:47:49.892 25423 read_cb:122: Closing Stream (0x7ffc922b6ee0): EOF (end of file)
    DEBUG 2018-09-18T09:47:49.892 25423 read_cb:122: Closing Stream (0x7ffc922b6d40): EOF (end of file)
    INFO  2018-09-18T09:47:49.893 25423 on_process_exit:385: exited: pid=25428 status=0 stoptime=0
    DEBUG 2018-09-18T09:47:49.894 25423 stream_close:85: closing Stream: 0x7ffc922b6d40
    DEBUG 2018-09-18T09:47:49.894 25423 stream_close:85: closing Stream: 0x7ffc922b6ee0
    DEBUG 2018-09-18T09:47:49.894 25423 UI: busy_stop
    DEBUG 2018-09-18T09:47:49.894 25423 UI: raw_line
    DEBUG 2018-09-18T09:47:49.894 25423 UI: raw_line (+4 times...)
    DEBUG 2018-09-18T09:47:49.894 25423 UI: grid_scroll
    DEBUG 2018-09-18T09:47:49.894 25423 UI: grid_scroll (+1 times...)
    DEBUG 2018-09-18T09:47:49.894 25423 UI: raw_line
    ERROR 2018-09-18T09:47:49.894 25423 wait_return:912: XXX wait_return
    DEBUG 2018-09-18T09:47:49.894 25423 UI: grid_cursor_goto
    DEBUG 2018-09-18T09:47:49.894 25423 UI: flush
    DEBUG 2018-09-18T09:47:49.895 25423 log_server_msg:695: RPC ->ch 1: [notify]   [2, "redraw", [["busy_stop", []], ["grid_line", [1, 13, 0, [["2", 0]]], [1, 13, 1, [["5", 0]]], [1, 13, 2, [["4", 0]]], [1, 13, 3, [["2", 0]]], [1, 13, 4, [["3", 0]]]], ["grid_scroll", [1, 9, 14, 0, 53, 1, 0], [1, 8, 14, 0, 53, 1, 0]], ["grid_line", [1, 13, 0, [["P", 23], ["r"], ["e"], ["s", 23, 2], [" "], ["E"], ["N"], ["T"], ["E"], ["R"], [" "], ["o"], ["r"], [" "], ["t"], ["y"], ["p"], ["e"], [" "], ["c"], ["o"], ["m", 23, 2], ["a"], ["n"], ["d"], [" "], ["t"], ["o"], [" "], ["c"], ["o"], ["n"], ["t"], ["i"], ["n"], ["u"], ["e"]]]], ["grid_cursor_goto", [1, 13, 39]]]]
    ERROR 2018-09-18T09:47:49.896 25423 input_poll:356: XXX blocking... events_enabled=0 empty=1
    DEBUG 2018-09-18T09:47:49.896 25423 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:110
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vgetc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1403
      safe_vgetc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1517
      wait_return at /home/vagrant/neovim/build/../src/nvim/message.c:913
      do_cmdline at ??:?
      nv_colon at /home/vagrant/neovim/build/../src/nvim/normal.c:4594
      normal_execute at /home/vagrant/neovim/build/../src/nvim/normal.c:1140
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:72
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?
    DEBUG 2018-09-18T09:47:53.386 25423 receive_msgpack:226: ch 1: parsing 19 bytes from msgpack Stream: 0x7ffb8caad820
    DEBUG 2018-09-18T09:47:53.386 25423 log_client_msg:732: RPC <-ch 1: [request]  [0, 2, "nvim_ui_detach", []]
    DEBUG 2018-09-18T09:47:53.386 25423 handle_request:354: RPC: scheduled nvim_ui_detach
    ERROR 2018-09-18T09:47:53.930 25423 input_poll:356: XXX blocking... events_enabled=0 empty=0
    DEBUG 2018-09-18T09:47:53.930 25423 input_poll:358: trace:
      log_callstack at /home/vagrant/neovim/build/../src/nvim/log.c:236
      input_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:360
      inbuf_poll at /home/vagrant/neovim/build/../src/nvim/os/input.c:383
      os_inchar at /home/vagrant/neovim/build/../src/nvim/os/input.c:120
      inchar at /home/vagrant/neovim/build/../src/nvim/getchar.c:2430
      vgetorpeek at /home/vagrant/neovim/build/../src/nvim/getchar.c:2263
      vgetc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1403
      safe_vgetc at /home/vagrant/neovim/build/../src/nvim/getchar.c:1517
      wait_return at /home/vagrant/neovim/build/../src/nvim/message.c:913
      do_cmdline at ??:?
      nv_colon at /home/vagrant/neovim/build/../src/nvim/normal.c:4594
      normal_execute at /home/vagrant/neovim/build/../src/nvim/normal.c:1140
      state_enter at /home/vagrant/neovim/build/../src/nvim/state.c:72
      normal_enter at /home/vagrant/neovim/build/../src/nvim/normal.c:467
      main at /home/vagrant/neovim/build/../src/nvim/main.c:567
      ?? ??:0
      _start at ??:?

justinmk committed:
WIP: multiqueue.c: more fixes? for size

bfredl commented on src/nvim/os/input.c:
```diff
@@ -88,9 +88,9 @@ static void create_cursorhold_event(void)
+  // TODO(justinmk): Cursorhold should be implemented as a timer set during the
   // `state_check` callback for the states where it can be triggered.
-  assert(!events_enabled || multiqueue_empty(main_loop.events));
+  assert(multiqueue_empty(main_loop.events));
```
Why is that? is K_EVENT regarded as a dismissive key in `wait_return`? Maybe it needs to be added to the list at `messages.c:954-966` of "allowed" state r keys.
